File: week-2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <ctype.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void) 
{
    // TODO: Get text to interpret reading level from
    string text = get_string("Text: ");

    // TODO: Get letters
    int letters = count_letters(text);

    // TODO: Get words
    int words = count_words(text);

    // TODO: Get sentences
    int sentences = count_sentences(text);

    // TODO: Calc averages & output grade level
    float avg_letters = (float) letters / (float) words * 100.0;
    float avg_sentences = (float) sentences / (float) words * 100.0;
    int level = round(0.0588 * avg_letters - 0.296 * avg_sentences - 15.8);

    // TODO: >= 16 = print "Grade 16+"
    if (level >= 16)
    {
        printf("Grade 16+\n");
    }
    // TODO: < 1 = print "Before Grade 1"
    else if (level < 1)
    {
        printf("Before Grade 1\n");
    }
    // TODO: Round to nearest int & print "Grade X"
    else
    {
        printf("Grade %i\n", level);
    }
}

int count_letters(string text)
{
    int lCount = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (isalpha(text[i]))
        {
            lCount++;
        }
    }
    return lCount;
}

int count_words(string text)
{
    int wCount = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (isspace(text[i]))
        {
            wCount++;
        }
    }
    return wCount + 1;
}

int count_sentences(string text)
{
    int sCount = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (text[i] == '.' || text[i] == '?' || text[i] == '!')
        {
            sCount++;
        }
    }
    return sCount;
}

File: week-6/sentimental-readability/readability.py
#!/usr/bin/env python3
from cs50 import get_string
import re


def main():
    # Get string of text from user for analysis
    text = get_string("Text: ")

    # Calculate num of letters (A-Z/a-z)
    letters = len(re.findall(r"[A-Za-z]", text))

    # Calculate num of words (any sequence of chars separated by spaces on
    # either end)
    words = len(re.findall(r"\s*\S+\s*", text))

    # Calculate num of sentences (.!?)
    sentences = len(re.findall(r"(\w*)[.!?]", text))

    # Calculate avg letters per 100 words
    avg_letters = letters / words * 100

    # Calculate avg sentences per 100 words
    avg_sentences = sentences / words * 100

    # Use Coleman-Liau formula to calculate reading level of input text
    reading_level = 0.0588 * avg_letters - 0.296 * avg_sentences - 15.8

    # Print grade level
    # If < 1, print "Before Grade 1"
    if reading_level < 1:
        print("Before Grade 1")
    # If 16+, print "16+" -- senior undergraduate reading level
    elif reading_level >= 16:
        print("Grade 16+")
    # Else round result to whole num for output
    else:
        print(f"Grade {reading_level:.0f}")


if __name__ == "__main__":
    main()
